File: golem-worker-executor-base/src/services/key_value.rs
// Copyright 2024-2025 Golem Cloud
//
// Licensed under the Golem Source License v1.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://license.golem.cloud/LICENSE
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::sync::Arc;

use anyhow::anyhow;
use async_trait::async_trait;
use bytes::Bytes;

use golem_common::model::AccountId;

use crate::storage::keyvalue::{
    KeyValueStorage, KeyValueStorageLabelledApi, KeyValueStorageNamespace,
};

/// Service implementing a persistent key-value store
#[async_trait]
pub trait KeyValueService: Send + Sync {
    async fn delete(
        &self,
        account_id: AccountId,
        bucket: String,
        key: String,
    ) -> anyhow::Result<()>;

    async fn delete_many(
        &self,
        account_id: AccountId,
        bucket: String,
        keys: Vec<String>,
    ) -> anyhow::Result<()>;

    async fn exists(
        &self,
        account_id: AccountId,
        bucket: String,
        key: String,
    ) -> anyhow::Result<bool>;

    async fn get(
        &self,
        account_id: AccountId,
        bucket: String,
        key: String,
    ) -> anyhow::Result<Option<Vec<u8>>>;

    async fn get_keys(&self, account_id: AccountId, bucket: String) -> anyhow::Result<Vec<String>>;

    async fn get_many(
        &self,
        account_id: AccountId,
        bucket: String,
        keys: Vec<String>,
    ) -> anyhow::Result<Vec<Option<Vec<u8>>>>;

    async fn set(
        &self,
        account_id: AccountId,
        bucket: String,
        key: String,
        outgoing_value: Vec<u8>,
    ) -> anyhow::Result<()>;

    async fn set_many(
        &self,
        account_id: AccountId,
        bucket: String,
        key_values: Vec<(String, Vec<u8>)>,
    ) -> anyhow::Result<()>;
}

#[derive(Clone, Debug)]
pub struct DefaultKeyValueService {
    key_value_storage: Arc<dyn KeyValueStorage + Send + Sync>,
}

impl DefaultKeyValueService {
    pub fn new(key_value_storage: Arc<dyn KeyValueStorage + Send + Sync>) -> Self {
        Self { key_value_storage }
    }
}

#[async_trait]
impl KeyValueService for DefaultKeyValueService {
    async fn delete(
        &self,
        account_id: AccountId,
        bucket: String,
        key: String,
    ) -> anyhow::Result<()> {
        self.key_value_storage
            .with("key_value", "delete")
            .del(
                KeyValueStorageNamespace::UserDefined { account_id, bucket },
                &key,
            )
            .await
            .map_err(|err| anyhow!(err))?;
        Ok(())
    }

    async fn delete_many(
        &self,
        account_id: AccountId,
        bucket: String,
        keys: Vec<String>,
    ) -> anyhow::Result<()> {
        self.key_value_storage
            .with("key_value", "delete_many")
            .del_many(
                KeyValueStorageNamespace::UserDefined { account_id, bucket },
                keys,
            )
            .await
            .map_err(|err| anyhow!(err))?;
        Ok(())
    }

    async fn exists(
        &self,
        account_id: AccountId,
        bucket: String,
        key: String,
    ) -> anyhow::Result<bool> {
        let exists: bool = self
            .key_value_storage
            .with("key_value", "exists")
            .exists(
                KeyValueStorageNamespace::UserDefined { account_id, bucket },
                &key,
            )
            .await
            .map_err(|err| anyhow!(err))?;
        Ok(exists)
    }

    async fn get(
        &self,
        account_id: AccountId,
        bucket: String,
        key: String,
    ) -> anyhow::Result<Option<Vec<u8>>> {
        let incoming_value: Option<Vec<u8>> = self
            .key_value_storage
            .with_entity("key_value", "get", "custom")
            .get_raw(
                KeyValueStorageNamespace::UserDefined { account_id, bucket },
                &key,
            )
            .await
            .map_err(|err| anyhow!(err))?
            .map(|bytes| bytes.to_vec());
        Ok(incoming_value)
    }

    async fn get_keys(&self, account_id: AccountId, bucket: String) -> anyhow::Result<Vec<String>> {
        let keys: Vec<String> = self
            .key_value_storage
            .with("key_value", "get_keys")
            .keys(KeyValueStorageNamespace::UserDefined { account_id, bucket })
            .await
            .map_err(|err| anyhow!(err))?;
        Ok(keys)
    }

    async fn get_many(
        &self,
        account_id: AccountId,
        bucket: String,
        keys: Vec<String>,
    ) -> anyhow::Result<Vec<Option<Vec<u8>>>> {
        let incoming_values: Vec<Option<Bytes>> = self
            .key_value_storage
            .with_entity("key_value", "get_many", "custom")
            .get_many_raw(
                KeyValueStorageNamespace::UserDefined { account_id, bucket },
                keys,
            )
            .await
            .map_err(|err| anyhow!(err))?;
        Ok(incoming_values
            .into_iter()
            .map(|mb| mb.map(|b| b.to_vec()))
            .collect())
    }

    async fn set(
        &self,
        account_id: AccountId,
        bucket: String,
        key: String,
        outgoing_value: Vec<u8>,
    ) -> anyhow::Result<()> {
        self.key_value_storage
            .with_entity("key_value", "set", "custom")
            .set_raw(
                KeyValueStorageNamespace::UserDefined { account_id, bucket },
                &key,
                &outgoing_value,
            )
            .await
            .map_err(|err| anyhow!(err))?;
        Ok(())
    }

    async fn set_many(
        &self,
        account_id: AccountId,
        bucket: String,
        key_values: Vec<(String, Vec<u8>)>,
    ) -> anyhow::Result<()> {
        let key_values: Vec<(&str, &[u8])> = key_values
            .iter()
            .map(|(key, value)| (key.as_str(), value.as_slice()))
            .collect();
        self.key_value_storage
            .with_entity("key_value", "set_many", "custom")
            .set_many_raw(
                KeyValueStorageNamespace::UserDefined { account_id, bucket },
                &key_values,
            )
            .await
            .map_err(|err| anyhow!(err))?;
        Ok(())
    }
}

File: sdks/scala/core/js/src/main/scala/golem/wasi/KeyValue.scala
/*
 * Copyright 2024-2026 John A. De Goes and the ZIO Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package golem.wasi

import golem.host.js._

import scala.scalajs.js
import scala.scalajs.js.annotation.JSImport
import scala.scalajs.js.typedarray.Uint8Array

/**
 * Scala.js facade for WASI keyvalue (`wasi:keyvalue@0.1.0`).
 *
 * Provides typed access to the keyvalue store via `Bucket`, `OutgoingValue`,
 * and `IncomingValue` resources. The Component Model JS binding unwraps
 * `result` types (throwing on `err`, returning `ok` directly) and maps `option`
 * to `undefined`.
 *
 * WIT interfaces:
 * {{{
 *   // wasi:keyvalue/types@0.1.0
 *   resource bucket { open-bucket: static func(name: string) -> result<bucket, error> }
 *   resource outgoing-value { new-outgoing-value: static func() -> outgoing-value; ... }
 *   resource incoming-value { incoming-value-consume-sync: func() -> result<list<u8>, error>; ... }
 *
 *   // wasi:keyvalue/eventual@0.1.0
 *   get: func(bucket, key) -> result<option<incoming-value>, error>
 *   set: func(bucket, key, outgoing-value) -> result<_, error>
 *   delete: func(bucket, key) -> result<_, error>
 *   exists: func(bucket, key) -> result<bool, error>
 *
 *   // wasi:keyvalue/eventual-batch@0.1.0
 *   get-many: func(bucket, keys) -> result<list<option<incoming-value>>, error>
 *   keys: func(bucket) -> result<list<key>, error>
 *   set-many: func(bucket, key-values) -> result<_, error>
 *   delete-many: func(bucket, keys) -> result<_, error>
 * }}}
 */
object KeyValue {

  // --- Native imports ---

  @js.native
  @JSImport("wasi:keyvalue/eventual@0.1.0", JSImport.Namespace)
  private object EventualModule extends js.Object {
    def get(bucket: js.Any, key: String): js.Any              = js.native
    def set(bucket: js.Any, key: String, value: js.Any): Unit = js.native
    def delete_(bucket: js.Any, key: String): Unit            = js.native
    def exists(bucket: js.Any, key: String): Boolean          = js.native
  }

  @js.native
  @JSImport("wasi:keyvalue/eventual-batch@0.1.0", JSImport.Namespace)
  private object EventualBatchModule extends js.Object {
    def getMany(bucket: js.Any, keys: js.Array[String]): js.Array[js.Any] = js.native
    def keys(bucket: js.Any): js.Array[String]                            = js.native
    def setMany(bucket: js.Any, keyValues: js.Array[js.Any]): Unit        = js.native
    def deleteMany(bucket: js.Any, keys: js.Array[String]): Unit          = js.native
  }

  @js.native
  @JSImport("wasi:keyvalue/types@0.1.0", "Bucket")
  private object KvBucketClass extends js.Object {
    def openBucket(name: String): js.Any = js.native
  }

  @js.native
  @JSImport("wasi:keyvalue/types@0.1.0", "OutgoingValue")
  private object KvOutgoingValueClass extends js.Object {
    def newOutgoingValue(): JsKvOutgoingValue = js.native
  }

  @js.native
  @JSImport("wasi:keyvalue/types@0.1.0", JSImport.Namespace)
  private object TypesModule extends js.Object

  @js.native
  @JSImport("wasi:keyvalue/wasi-keyvalue-error@0.1.0", JSImport.Namespace)
  private object ErrorModule extends js.Object

  // --- Bucket resource ---

  final class Bucket private[KeyValue] (private[golem] val underlying: js.Any) {

    def get(key: String): Option[Array[Byte]] = {
      val result = EventualModule.get(underlying, key)
      if (js.isUndefined(result) || result == null) None
      else {
        val iv = new IncomingValue(result.asInstanceOf[JsKvIncomingValue])
        Some(iv.consumeSync())
      }
    }

    def set(key: String, value: Array[Byte]): Unit = {
      val ov = OutgoingValue.create()
      ov.writeSync(value)
      EventualModule.set(underlying, key, ov.underlying)
    }

    def delete(key: String): Unit =
      EventualModule.delete_(underlying, key)

    def exists(key: String): Boolean =
      EventualModule.exists(underlying, key)

    def keys(): List[String] =
      EventualBatchModule.keys(underlying).toList

    def getMany(keys: List[String]): List[Option[Array[Byte]]] = {
      val arr = js.Array(keys: _*)
      EventualBatchModule.getMany(underlying, arr).toList.map { result =>
        if (js.isUndefined(result) || result == null) None
        else Some(new IncomingValue(result.asInstanceOf[JsKvIncomingValue]).consumeSync())
      }
    }

    def deleteMany(keys: List[String]): Unit =
      EventualBatchModule.deleteMany(underlying, js.Array(keys: _*))
  }

  object Bucket {
    def open(name: String): Bucket = {
      val raw = KvBucketClass.openBucket(name)
      new Bucket(raw)
    }
  }

  // --- OutgoingValue resource ---

  final class OutgoingValue private[KeyValue] (private[golem] val underlying: JsKvOutgoingValue) {

    def writeSync(data: Array[Byte]): Unit = {
      val jsArr = js.Array[Short]()
      data.foreach(b => jsArr.push((b.toInt & 0xff).toShort))
      val bytes = new Uint8Array(jsArr.asInstanceOf[js.Iterable[Short]])
      underlying.outgoingValueWriteBodySync(bytes)
    }
  }

  object OutgoingValue {
    def create(): OutgoingValue = {
      val raw = KvOutgoingValueClass.newOutgoingValue()
      new OutgoingValue(raw)
    }
  }

  // --- IncomingValue resource ---

  final class IncomingValue private[KeyValue] (private[golem] val underlying: JsKvIncomingValue) {

    def consumeSync(): Array[Byte] = {
      val arr   = underlying.incomingValueConsumeSync()
      val bytes = new Array[Byte](arr.length)
      var i     = 0
      while (i < arr.length) {
        bytes(i) = arr(i).toByte
        i += 1
      }
      bytes
    }

    def size(): Long = {
      val sizeBigInt = BigInt(underlying.incomingValueSize().toString)
      if (!sizeBigInt.isValidLong)
        throw new IllegalArgumentException(s"Incoming value size $sizeBigInt does not fit into a Long")
      sizeBigInt.toLong
    }
  }

  // --- Raw access (for forward compatibility) ---

  def eventualRaw: Any      = EventualModule
  def eventualBatchRaw: Any = EventualBatchModule
  def typesRaw: Any         = TypesModule
  def errorRaw: Any         = ErrorModule
}
